File: iosApp/iosAppTests/Pages/StopDetails/StopDetailsNoTripCardTests.swift
//
//  StopDetailsNoTripCardTests.swift
//  iosAppTests
//
//  Created by esimon on 1/3/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

@testable import iosApp
import Shared
import SwiftUI
import ViewInspector
import XCTest

final class StopDetailsNoTripCardTests: XCTestCase {
    override func setUp() {
        executionTimeAllowance = 60
    }

    func testSubwayEarlyMorning() throws {
        let serviceStart = EasternTimeInstant(year: 2025, month: .november, day: 17, hour: 10, minute: 25, second: 0)
        let sut = StopDetailsNoTripCard(
            status: UpcomingFormat.NoTripsFormatSubwayEarlyMorning(scheduledTime: serviceStart),
            accentColor: Color.text,
            directionLabel: "Forest Hills",
            routeType: .heavyRail,
            now: .now()
        ).withFixedSettings([:])
        XCTAssertNotNil(try sut.inspect().find(imageName: "sunrise"))
        XCTAssertNotNil(try sut.inspect().find(text: "Good morning!"))
        XCTAssertNotNil(try sut.inspect().find(ViewType.Divider.self))
        XCTAssertNotNil(try sut.inspect().find(
            textWhere: { value, _ in
                try #/The first Forest Hills train is scheduled to arrive at 10:25\sAM. We don’t have predictions to show you yet, but they’ll appear here closer to the scheduled time./#
                    .wholeMatch(in: value) != nil
            }
        ))
    }

    func testPredictionsUnavailable() throws {
        let sut = StopDetailsNoTripCard(
            status: UpcomingFormat.NoTripsFormatPredictionsUnavailable(),
            accentColor: Color.text,
            directionLabel: "Forest Hills",
            routeType: .bus,
            now: .now()
        ).withFixedSettings([:])
        XCTAssertNotNil(try sut.inspect().find(imageName: "live-data-slash"))
        XCTAssertNotNil(try sut.inspect().find(text: "Predictions unavailable"))
        XCTAssertNotNil(try sut.inspect().find(ViewType.Divider.self))
        XCTAssertNotNil(try sut.inspect().find(
            text: "Service is running, but predicted arrival times aren’t available." +
                " Check the map to see where buses are right now."
        ))
    }

    func testServiceEnded() throws {
        let objects = ObjectCollectionBuilder()
        let now = EasternTimeInstant.now()
        let schedule = objects.schedule { $0.departureTime = .init(
            date: now.local.date.plus(days: 1),
            time: .init(hour: 9, minute: 15, second: 0, nanosecond: 0)
        ) }
        let sut = StopDetailsNoTripCard(
            status: UpcomingFormat.NoTripsFormatServiceEndedToday(),
            accentColor: Color.text,
            directionLabel: "Winthrop",
            routeType: .ferry,
            now: now,
            nextScheduleResponse: .init(nextSchedule: schedule)
        ).withFixedSettings([:])
        XCTAssertNotNil(try sut.inspect().find(imageName: "mode-ferry-slash"))
        XCTAssertNotNil(try sut.inspect().find(text: "Service ended"))
        XCTAssertNotNil(try sut.inspect().find(textWhere: { text, _ in
            text.wholeMatch(of: #/Next trip at 9:15\sAM tomorrow/#) != nil
        }))
    }

    func testNoSchedulesToday() throws {
        let objects = ObjectCollectionBuilder()
        let now = EasternTimeInstant.now()
        let schedule = objects.schedule { $0.departureTime = .init(
            date: now.local.date,
            time: .init(hour: 9, minute: 15, second: 0, nanosecond: 0)
        ) }
        let sut = StopDetailsNoTripCard(
            status: UpcomingFormat.NoTripsFormatNoSchedulesToday(),
            accentColor: Color.text,
            directionLabel: "Fitchburg",
            routeType: .commuterRail,
            now: now,
            nextScheduleResponse: .init(nextSchedule: schedule)
        ).withFixedSettings([:])
        XCTAssertNotNil(try sut.inspect().find(imageName: "mode-cr-slash"))
        XCTAssertNotNil(try sut.inspect().find(text: "No service today"))
        XCTAssertNotNil(try sut.inspect().find(textWhere: { text, _ in
            text.wholeMatch(of: #/Next trip at 9:15\sAM/#) != nil
        }))
    }
}

File: androidApp/src/androidTest/java/com/mbta/tid/mbta_app/android/stopDetails/StopDetailsNoTripCardTests.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.test.junit4.createComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import com.mbta.tid.mbta_app.android.loadKoinMocks
import com.mbta.tid.mbta_app.android.testUtils.assertCanBeDisplayed
import com.mbta.tid.mbta_app.android.testUtils.hasTextMatching
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.UpcomingFormat
import com.mbta.tid.mbta_app.model.response.NextScheduleResponse
import com.mbta.tid.mbta_app.repositories.MockSettingsRepository
import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.LocalTime
import kotlinx.datetime.Month
import kotlinx.datetime.plus
import org.junit.Rule
import org.junit.Test

class StopDetailsNoTripCardTests {

    @get:Rule val composeTestRule = createComposeRule()

    @Test
    fun testSubwayEarlyMorning() {
        loadKoinMocks()
        composeTestRule.setContent {
            StopDetailsNoTripCard(
                status =
                    UpcomingFormat.NoTripsFormat.SubwayEarlyMorning(
                        EasternTimeInstant(2025, Month.NOVEMBER, 17, 9, 44)
                    ),
                accentColor = Color.Black,
                directionLabel = "Forest Hills",
                routeType = RouteType.HEAVY_RAIL,
                now = EasternTimeInstant.now(),
                nextScheduleResponse = null,
            )
        }

        composeTestRule.onNodeWithTag("sunrise").assertCanBeDisplayed()
        composeTestRule.onNodeWithText("Good morning!").assertCanBeDisplayed()
        composeTestRule
            .onNode(
                hasTextMatching(
                    Regex(
                        "^The first Forest Hills train is scheduled to arrive at 9:44\\sAM. We don’t have predictions to show you yet, but they’ll appear here closer to the scheduled time.$",
                        RegexOption.IGNORE_CASE,
                    )
                )
            )
            .assertCanBeDisplayed()
    }

    @Test
    fun testPredictionsUnavailable() {
        loadKoinMocks()
        composeTestRule.setContent {
            StopDetailsNoTripCard(
                status = UpcomingFormat.NoTripsFormat.PredictionsUnavailable,
                accentColor = Color.Black,
                directionLabel = "Forest Hills",
                routeType = RouteType.BUS,
                now = EasternTimeInstant.now(),
                nextScheduleResponse = null,
            )
        }

        composeTestRule.onNodeWithTag("live_data_slash").assertCanBeDisplayed()
        composeTestRule.onNodeWithText("Predictions unavailable").assertCanBeDisplayed()
        composeTestRule
            .onNodeWithText(
                "Service is running, but predicted arrival times aren’t available." +
                    " Check the map to see where buses are right now."
            )
            .assertCanBeDisplayed()
    }

    @Test
    fun testPredictionsUnavailableHideMaps() {
        loadKoinMocks { settings = MockSettingsRepository(mapOf(Settings.HideMaps to true)) }

        composeTestRule.setContent {
            StopDetailsNoTripCard(
                status = UpcomingFormat.NoTripsFormat.PredictionsUnavailable,
                accentColor = Color.Black,
                directionLabel = "Forest Hills",
                routeType = RouteType.BUS,
                now = EasternTimeInstant.now(),
                nextScheduleResponse = null,
            )
        }

        composeTestRule
            .onNodeWithText(
                "The map shows where buses on this route currently are.",
                substring = true,
            )
            .assertDoesNotExist()
    }

    @Test
    fun testServiceEnded() {
        val objects = ObjectCollectionBuilder()
        val now = EasternTimeInstant.now()
        val schedule =
            objects.schedule {
                departureTime =
                    EasternTimeInstant(
                        now.local.date.plus(DatePeriod(days = 1)),
                        time = LocalTime(hour = 9, minute = 15),
                    )
            }
        loadKoinMocks()
        composeTestRule.setContent {
            StopDetailsNoTripCard(
                status = UpcomingFormat.NoTripsFormat.ServiceEndedToday,
                accentColor = Color.Black,
                directionLabel = "Winthrop",
                routeType = RouteType.FERRY,
                now = now,
                nextScheduleResponse = NextScheduleResponse(schedule),
            )
        }
        composeTestRule.onNodeWithTag("route_slash_icon").assertCanBeDisplayed()
        composeTestRule.onNodeWithText("Service ended").assertCanBeDisplayed()
        composeTestRule
            .onNode(
                hasTextMatching(Regex("^Next trip at 9:15\\sAM tomorrow$", RegexOption.IGNORE_CASE))
            )
            .assertCanBeDisplayed()
    }

    @Test
    fun testNoSchedulesToday() {
        val objects = ObjectCollectionBuilder()
        val now = EasternTimeInstant.now()
        val schedule =
            objects.schedule {
                departureTime =
                    EasternTimeInstant(now.local.date, time = LocalTime(hour = 9, minute = 15))
            }
        loadKoinMocks()
        composeTestRule.setContent {
            StopDetailsNoTripCard(
                status = UpcomingFormat.NoTripsFormat.NoSchedulesToday,
                accentColor = Color.Black,
                directionLabel = "Fitchburg",
                routeType = RouteType.COMMUTER_RAIL,
                now = now,
                nextScheduleResponse = NextScheduleResponse(schedule),
            )
        }
        composeTestRule.onNodeWithTag("route_slash_icon").assertCanBeDisplayed()
        composeTestRule.onNodeWithText("No service today").assertCanBeDisplayed()
        composeTestRule
            .onNode(hasTextMatching(Regex("^Next trip at 9:15\\sAM$", RegexOption.IGNORE_CASE)))
            .assertCanBeDisplayed()
    }
}
